proposed/swift-interop.md

```diff
@@ -0,0 +1,132 @@
+
+The "SwiftAsync" calling convention has an additional "async context" register and specific behaviors expected around how the Swift async state machine is called. Full investigation into the differences needed to implement it are still TBD.
+
+##### Self register
+
+We have two options for supporting the Self register in the calling convention:
+
+1. Use an attribute like `[SwiftSelf]` on a parameter to specify it should go in the self register.
+2. Specify that combining the `Swift` calling convention with the `MemberFunction` calling convention means that the first argument is the `self` argument.
+
+The first option seems like a natural fit, but there is one significant limitation: Attributes cannot be used in function pointers. Function pointers are a vital scenario for us to support as we want to support virtual method tables as they are used in scenarios like protocol witness tables. The second option is a natural fit as the `MemberFunction` calling convention combined with the various C-based calling conventions specifies that there is a `this` argument as the first argument. Defining `Swift` + `MemberFunction` to imply/require the `self` argument is a perfect conceptual extension of the existing model.
+
+###### Error register
+
+We have many options for handling the error register in the Swift calling convention:
+
+1. Use an attribute like `[SwiftError]` on a by-ref or `out` parameter to indicate the error parameter.
+2. Use a special type like `SwiftError<T>` on a by-ref or `out` parameter to indicate the error parameter.
+3. Use a special return type `SwiftReturn<T, TError>` to indicate the error parameter and combine it with the return value.
+4. Use special helper functions `Marshal.Get/SetLastSwiftError` to get and set the last Swift error. Our various compilers will automatically emit a call to `SetLastSwiftError` from Swift functions and emit a call to `GetLastSwiftError` in the epilog of `UnmanagedCallersOnly` methods. The projection generation will emit calls to `Marshal.Get/SetLastSwiftError` to get and set these values to and from exceptions.
+
+We have a prototype that uses the first option; however, using an attribute has the same limitations as the attribute option for the self register. As a result, we should consider an alternative option.
+
+Options 2 and 3 have similar characteristics: They both express that the to-be-called Swift function uses the Error Register in the signature and they both require signature manipulation in the JIT/AOT compilers. Option 2 is likely cheaper, as the argument with `ref/out SwiftError<T>` type can easily be associated with the error register. With Option 2 though, we'd have to determine the behavior if multiple parameters used the `SwiftError` type (likely throw an exception). Option 3 provides a more functional-programming-style API that is more intuitive, but likely more expensive to implement.
```

Comment:
> With Option 2 though, we'd have to determine the behavior if multiple parameters used the `SwiftError` type (likely throw an exception).

What are implications if multiple parameters used the `SwiftError` type?

Reply:
We'd likely want to throw an `InvalidProgramException` as that would not be a valid signature.